File: flink-sample-java/src/main/java/CoProcessFunctionTimers.java
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import udf.ReadingFilter;
import util.SensorReading;
import util.SensorSource;

public class CoProcessFunctionTimers {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setStreamTimeCharacteristic(TimeCharacteristic.ProcessingTime);

        DataStream<Tuple2<String, Long>> filterSwitchs = env
                .fromElements(Tuple2.of("sensor_2", 10_000L), Tuple2.of("sensor_7", 60_000L));
        DataStream<SensorReading> readings = env.addSource(new SensorSource());

        DataStream<SensorReading> forwardedReadings = readings
                .connect(filterSwitchs)
                .keyBy(r -> r.id, s -> s.f0)
                .process(new ReadingFilter());

        forwardedReadings.print();
        env.execute();
    }
}

File: flink-datastream-1.7.1/src/main/scala/scalalang/CoProcessFunctionTimers.scala
package scalalang

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.scala._
import scalalang.function.ReadingFilter
import scalalang.source.SensorSource

object CoProcessFunctionTimers {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    //处理时间
    env.setStreamTimeCharacteristic(TimeCharacteristic.ProcessingTime)
    //两条输入流
    val filterSwitches = env
      .fromCollection(Seq(("sensor_2", 10 * 1000L), ("sensor_7", 60 * 1000L)))
    val readings = env
      .addSource(new SensorSource)
    //关联、键值、处理
    val forwardedReadings = readings
      .connect(filterSwitches)
      .keyBy(_.id, _._1)
      .process(new ReadingFilter)
    //输出流
    forwardedReadings.print()
    env.execute()
  }
}
